cnd/src/network/protocol.rs

```diff
@@ -0,0 +1,160 @@
+use futures::{future, AsyncRead, AsyncWrite, Future};
+use libp2p::core::{upgrade, InboundUpgrade, OutboundUpgrade, UpgradeInfo};
+use serde::{Deserialize, Serialize};
+use std::{iter, pin::Pin};
+use tracing::trace;
+
+const LENGTH: usize = 32;
+
+lazy_static::lazy_static! {
+    static ref JSON_LENGTH: usize = {
+        let info = SecretHashInfo {
+            swap_id: [0; LENGTH],
+            secret_hash: [0; LENGTH],
+        };
+
+        let serialized = serde_json::to_vec(&info).unwrap();
+
+        serialized.len()
+    };
+}
+
+/// The secret hash sharing protocol works in the following way:
+///
+/// - Dialer (Alice) writes the `SecretHashInfo` to the substream.
+/// - Listener (Bob) reads the `SecretHashInfo` from the substream.
+
+/// Represents a prototype for an upgrade to handle the secret hash sharing
+/// protocol.
+#[derive(Debug, Clone)]
+pub struct SecretHashProtocolConfig {
+    info: Option<SecretHashInfo>,
+}
+
+impl SecretHashProtocolConfig {
+    /// Sender sends the secret hash `info` to the peer.
+    pub fn sender(info: SecretHashInfo) -> Self {
+        SecretHashProtocolConfig { info: Some(info) }
+    }
+
+    /// Receiver receives the secret hash from the peer.
+    pub fn receiver() -> Self {
+        SecretHashProtocolConfig { info: None }
+    }
+}
+
+/// Data sent to peer in secret hash protocol.
+#[derive(Clone, Deserialize, Debug, Serialize)]
+pub struct SecretHashInfo {
+    swap_id: [u8; LENGTH],
+    secret_hash: [u8; LENGTH],
```

Comment:
These two fields just happen to have the same length but that is not semantically related. Using the same const for that feels wrong to me.

Reply:
And its wrong altogether :)  one of them is 16 bytes and one is 32.  I've fixed this, sorry for leaving an old version open for review.  Same goes for the `lazy_static` comment above.